### PR TITLE
A more consistent kernel launch parameter syntax

### PR DESCRIPTION
**Pull request by chudur-budur**
Addressing issues #867 and #871. This might also fix issue #886 
Also related to issues #790 and #766

- [x] Have you provided a meaningful PR description?
- [x] Have you added a test, reproducer or referred to an issue with a reproducer?
- [x] Have you tested your changes locally for CPU and GPU devices?
- [x] Have you made sure that new changes do not introduce compiler warnings?
- [x] If this PR is a work in progress, are you filing the PR as a draft?

### CODE REVIEW

File: numba_dpex/core/kernel_interface/dispatcher.py [hunk 1]
```diff
@@ -468,52 +522,56 @@ def __getitem__(self, args):
...
-                elif len(gr) != 0 and all(isinstance(v, int) for v in gr):
-                    self._global_range = list(gr)
-                else:
-                    raise IllegalRangeValueError(kernel_name=self.kernel_name)
 
-                if isinstance(lr, int):
-                    self._local_range = [lr]
-                elif isinstance(lr, list) and len(lr) == 0:
-                    # deprecation warning
+        if isinstance(args, Ranges):
```
Comment (diptorupd) [inline numba_dpex/core/kernel_interface/dispatcher.py:526]: Instead of a single class, my preference is to have two classes `Range` and NdRange`. I think the `Ranges` class can be renamed to `NdRange` with the `local_range` arg as required. A similar class with only one parameter will represent `Range`.

File: numba_dpex/core/kernel_interface/dispatcher.py [hunk 2]
```diff
@@ -468,52 +522,56 @@ def __getitem__(self, args):
...
+                isinstance(args, tuple)
+                and len(args) == 2
+                and isinstance(args[0], int)
+                and isinstance(args[1], int)
+            ):
+                warn(
+                    "Ambiguous kernel launch paramters. "
+                    + "If your data have dimensions > 1, "
+                    + "include a default/empty local_range. "
+                    + "i.e. <function>[(M,N), numba_dpex.DEFAULT_LOCAL_RANGE](<params>), "
```
Comment (diptorupd) [inline numba_dpex/core/kernel_interface/dispatcher.py:540]: The message is OK in it self. But, if we intend to totally deprecate passing tupples or lists to `__getitem__` let us tell the users about the new Range and NdRange classes.

File: numba_dpex/core/kernel_interface/dispatcher.py [hunk 3]
```diff
@@ -468,52 +522,56 @@ def __getitem__(self, args):
...
 
-                if isinstance(lr, int):
-                    self._local_range = [lr]
-                elif isinstance(lr, list) and len(lr) == 0:
-                    # deprecation warning
+        if isinstance(args, Ranges):
+            self._global_range = list(args.global_range)[::-1]
+            self._local_range = list(args.local_range)[::-1]
+        else:
+            if (
```
Comment (diptorupd) [inline numba_dpex/core/kernel_interface/dispatcher.py:530]: Print a deprecation notice right here asking users to start using `Range` and `NdRange` classes.

File: numba_dpex/core/kernel_interface/dispatcher.py [hunk 4]
```diff
@@ -441,6 +445,56 @@ def _determine_kernel_launch_queue(self, args, argtypes):
             else:
                 raise ExecutionQueueInferenceError(self.kernel_name)
 
+    def _raise_invalid_kernel_enqueue_args(self):
+        error_message = (
+            "Incorrect number of arguments for enqueuing numba_dpex.kernel. "
+            "Usage: device_env, global size, local size. "
```
Comment (diptorupd) [inline numba_dpex/core/kernel_interface/dispatcher.py:451]: The error message is wrong and seems to have been not updated for the last two years even as we made many API changes!! Please do not use this old error message.

File: numba_dpex/core/kernel_interface/dispatcher.py [hunk 5]
```diff
@@ -468,52 +522,56 @@ def __getitem__(self, args):
...
+                return self
+
+            if not isinstance(args, Iterable):
+                args = [args]
+
+            ls = None
+            nargs = len(args)
+            # Check if the kernel enquing arguments are sane
+            if nargs < 1 or nargs > 2:
+                self._raise_invalid_kernel_enqueue_args()
```
Comment (diptorupd) [inline numba_dpex/core/kernel_interface/dispatcher.py:555]: Directly raise `InvalidKernelLaunchArgsError` here the old `_raise_invalid_kernel_enqueue_args` has a totally wrong error message.

File: numba_dpex/core/kernel_interface/dispatcher.py [hunk 6]
```diff
@@ -468,52 +522,56 @@ def __getitem__(self, args):
...
+
+            gs = self._ensure_valid_work_item_grid(args[0])
+            # If the optional local size argument is provided
+            if nargs == 2:
+                if args[1] != []:
+                    ls = self._ensure_valid_work_group_size(args[1], gs)
```
Comment (diptorupd) [inline numba_dpex/core/kernel_interface/dispatcher.py:561]: No need to call `_ensure_valid_work_group_size`. The function can be removed altogether. We care checking ranges for validity in the new `_check_ranges` function.

File: numba_dpex/examples/kernel/vector_sum2D.py [hunk 7]
```diff
@@ -22,7 +22,7 @@ def data_parallel_sum(a, b, c):
 
 
 def driver(a, b, c, global_size):
-    data_parallel_sum[global_size](a, b, c)
+    data_parallel_sum[global_size, dpex.DEFAULT_LOCAL_SIZE](a, b, c)
```
Comment (diptorupd) [inline numba_dpex/examples/kernel/vector_sum2D.py:25]: Ok. But, how about demonstrating the use of the new and likely future method of using Range classes?

File: numba_dpex/core/kernel_interface/dispatcher.py (hunk 1 above)
Comment (chudur-budur) [reply numba_dpex/core/kernel_interface/dispatcher.py]: Done.

File: numba_dpex/core/kernel_interface/dispatcher.py (hunk 2 above)
Comment (chudur-budur) [reply numba_dpex/core/kernel_interface/dispatcher.py:502]: Done.

File: numba_dpex/core/kernel_interface/dispatcher.py (hunk 3 above)
Comment (chudur-budur) [reply numba_dpex/core/kernel_interface/dispatcher.py:492]: Done

File: numba_dpex/core/kernel_interface/dispatcher.py (hunk 4 above)
Comment (chudur-budur) [reply numba_dpex/core/kernel_interface/dispatcher.py]: Fixed.

File: numba_dpex/core/kernel_interface/dispatcher.py (hunk 5 above)
Comment (chudur-budur) [reply numba_dpex/core/kernel_interface/dispatcher.py]: Old functions are removed.

File: numba_dpex/core/kernel_interface/dispatcher.py (hunk 6 above)
Comment (chudur-budur) [reply numba_dpex/core/kernel_interface/dispatcher.py]: Old functions are removed.

File: numba_dpex/examples/kernel/vector_sum2D.py (hunk 7 above)
Comment (chudur-budur) [reply numba_dpex/examples/kernel/vector_sum2D.py:26]: A new driver function is added to showcase the new way of specifying kernel lauch params.

File: numba_dpex/core/kernel_interface/dispatcher.py [hunk 8]
```diff
@@ -468,51 +470,79 @@ def __getitem__(self, args):
...
-                if isinstance(gr, int):
-                    self._global_range = [gr]
-                elif len(gr) != 0 and all(isinstance(v, int) for v in gr):
-                    self._global_range = list(gr)
-                else:
-                    raise IllegalRangeValueError(kernel_name=self.kernel_name)
+        warn(
+            """
+            The current syntax for specification of kernel lauch parameters
+            will be deprecated soon. Users are requested to do it through
```
Comment (diptorupd) [inline numba_dpex/core/kernel_interface/dispatcher.py:476]:
```suggestion
            is deprecated. Users should set the kernel launch arguments through
```

Comment (diptorupd) [reply numba_dpex/core/kernel_interface/dispatcher.py:476]: Moreover, why should I get the deprecation warning if I am using the right API. The warning should be moved into the else block.

File: numba_dpex/core/kernel_interface/dispatcher.py [hunk 9]
```diff
@@ -468,51 +470,79 @@ def __getitem__(self, args):
...
+            Range/NdRange classes.
+                Example:
+                    from numba_dpex.core.kernel_interface.utils import NdRange
+                    ranges = NdRange((8,8), (2,2)) # 2D global and local ranges
+                    <function>[ranges](<parameters>)
+            """,
+            DeprecationWarning,
+            stacklevel=2,
+        )
+        if isinstance(args, NdRange):
```
Comment (diptorupd) [inline numba_dpex/core/kernel_interface/dispatcher.py:486]: The change does not match SYCL semantics. The local range for an NdRange is not optional. We really need two classes `Range` and `NdRange`. There is a difference in how a kernel is submitted based on  whether a range or an ndrange is provided. 

We have to be clear and explicit about the usage. When using NdRange class it should without ambiguity indicate that the kernel will be submitted as an ndrange kernel.

Comment (chudur-budur) [reply numba_dpex/core/kernel_interface/dispatcher.py:486]: @diptorupd Right now, the correct syntax for kernel parameters are like one of these:

```python
ranges = NdRange((2,2,2),(1,1,1))
func[ranges](a,b,c)
```

or 

```python
ranges = NdRange((2,2,2))
func[ranges](a,b,c)
```

or

```python
global_range = Range(2,2,2)
local_range = Range(1,1,1)
ranges = NdRange(global_range, local_range)
func[ranges](a,b,c)
```

or

```python
global_range = Range(2,2,2)
ranges = NdRange(global_range)
func[ranges](a,b,c)
```

If these are not correct, could you give me an example of a correct syntax?

File: numba_dpex/core/kernel_interface/dispatcher.py (hunk 8 above)
Comment (chudur-budur) [reply numba_dpex/core/kernel_interface/dispatcher.py:476]: Fixed.

File: numba_dpex/core/kernel_interface/dispatcher.py (hunk 9 above)
Comment (diptorupd) [reply numba_dpex/core/kernel_interface/dispatcher.py:486]: In the above examples, (1) and (3) is what we should be support.
(2) and (4) we should not.

That is, `NdRange` should require both `global_range` and `local_range` to be set. We should not allow `NdRange` where `local_range` is `None`.

Comment (chudur-budur) [reply numba_dpex/core/kernel_interface/dispatcher.py:486]: > We should not allow `NdRange` where `local_range` is `None`.

But there are many examples where we don't know the `local_range` and therefore we use `numba_dpex.DEFAULT_LOCAL_SIZE` (which is basically a `None`). For example, what do we do in the `sum2D.py` in the `examples` folder?

Comment (chudur-budur) [reply numba_dpex/core/kernel_interface/dispatcher.py]: Fixed.

File: numba_dpex/config.py [hunk 10]
```diff
@@ -99,7 +99,7 @@ def __getattr__(name):
 ENABLE_CACHE = _readenv("NUMBA_DPEX_ENABLE_CACHE", int, 1)
 # Capacity of the cache, execute it like:
 #   NUMBA_DPEX_CACHE_SIZE=20 python <code>
-CACHE_SIZE = _readenv("NUMBA_DPEX_CACHE_SIZE", int, 10)
+CACHE_SIZE = _readenv("NUMBA_DPEX_CACHE_SIZE", int, 128)
```
Comment (diptorupd) [inline numba_dpex/config.py:102]: Why the change?

**Review by diptorupd — APPROVED** [review]
LGTM! Thank you @chudur-budur
